### DISCUSSION TITLE
Recognition in Character Level

### DISCUSSION THREAD

**Discussion by dhea1323** [mindee/doctr#950] [Q&A]
Hi,

doctr is the best OCR provided for free. The prediction results are also very impressive. I want to know the recognition stage, how does doctr recognition in character level? Would you like to explain the process?

Thank you

**Comment by frgfm** [comment]
Hi @dhea1323 :wave: 

Thanks for asking! I'll try to work my way to a rather exhaustive but brief answer. docTR uses three types of text recognition models:
- convolutional recurrent-based models (like CRNN): given an image, we resize it to a wide crop and extract visual features using convolutions. More or so, from (C, H, W) we get down to something like (C', 1, L). Then we used recurrent and linear layers to turn (C', 1, L) into (L, K) where K would be the number of output character class. Finally, we align the predictions using the CTC (I highly suggest reading this amazing [article](https://distill.pub/2017/ctc/) about it)
- Attention-based model (like SAR): instead of using a plain LSTM + linear combination, we use self-attention encoder & decoders to create the (L, K) shaped tensor that will be forwarded to the CTC decoder :)
- transformer-based model (like MASTER): finally, instead of the encoder/decoder blocks, we use a [Transformer](https://jalammar.github.io/illustrated-transformer/) Decoder layer.

If you're looking for more information about the implemented models in docTR, you will find the link to all the papers for each model in the [documentation](https://mindee.github.io/doctr/modules/models.html#doctr-models-recognition)!

I hope this helps, let me know if you have any questions :)

**Comment by AakashKumarNain** [comment]
@frgfm I have a follow-up question on this one. Let's say if the output of the feature extractor (CNN) is `(None, 1, L, C)` (assume NHWC format), we apply recurrent layers-dense layers to turn it into `(None, L, K)` where `K` is the number of output character class. How does doctr handle the cases where timesteps `L` << label length? For example if the network outputs `(None, 5, 11)` Here we have 5 timesteps but what if the length of true label is greater than 5? Wouldn't that be a problem for CTC?

**Reply by frgfm** [reply]
Hello @AakashKumarNain :wave:

For that, I would recommend considering characters included in existing supported vocabs. If that's the case, you can just remove the nodes that you don't want on your own (or simply remap the output characters)

If that's not the case, you have to finetune the final layers with your vocabulary :smile: 

Hope this clarifies things, let me know if you have more questions!

**Reply by AakashKumarNain** [reply]
Thank you for the response @frgfm 

> For that, I would recommend considering characters included in existing supported vocabs. If that's the case, you can just remove the nodes that you don't want on your own (or simply remap the output characters)

Can you please elaborate this a bit? All the characters in my dataset are also in the vocab. The problem is that the ground truth contains more characters than the length of the timesteps in the final layer. In this case, there are not enough transition steps, and designing a matrix would "ideally" be impossible. Any recommended workaround for this scenario?

**Reply by frgfm** [reply]
But what do you mean by "All the characters in my dataset are also in the vocab"? Was the model sized and trained with the specific characters you mention?

Your point about "The problem is that the ground truth contains more characters than the length of the timesteps in the final layer." is relevant because it assumes that we're expanding / upsampling data somewhere. But I think you're missing a point:
- with (C', 1, L) as input
- and (L, K) as output
there are axis swapes operating here. So even if L < your number of characters, it's not relevant. The layer making that transform will more or so map C' to K, not the axis of size L which is preserved.

If I'm missing your point here, please illustrate with examples :pray: But otherwise, I think you'll be fine with your characters and only need to filter/map the output characters :smiley:

**Reply by AakashKumarNain** [reply]
I got it what you are saying but just for the sake of clarity I will take an extreme case to make sure we both are on the same page regarding the timestep length. Assume everything in NHWC format, let's say you have images of size, say `(32 x 100 x 3)`, and the characters are the English alphabet . Let's break down the network into three parts: feature extractor, RNNs, output layer/s.

```
(1, 32, 100, 3)
            |
            |
(1, 1, 7, 512) # output of the feature extractor
            |
            |
  (1,  7, 256)   # RNN output
            |
            |
   (1,  7,  27)   # output of the last linear layer        
```

Let's say the corresponding label to this image is `comments section`. Are you saying that the time step length `7` in the output layer doesn't effect the CTC loss/predictions in this case because of the transform?

**Reply by frgfm** [reply]
More or so: here either the 7 will be upsampled, or (the option I recommend) you choose a better image size for your characters. In your example, 100 pixels are supposed to include "comments section" which is 16 characters including the blank space. For comparison, MNIST takes 32 pixels and it's only for digits :sweat_smile: 

So yes, the network is designed so that if your image is 100 pixels wide, it will be hard to distinguish more than a dozen characters. However, the assumption in text recognition is that the image doesn't have any blanks. So your image should be split in 2: one for "comments", one for "section". I hope that clears your interrogation :)

**Reply by AakashKumarNain** [reply]
> However, the assumption in text recognition is that the image doesn't have any blanks.

Yeah I am ware of that. I just saw it and I wanted to write `comments-section` but somehow it became a typo. Regarding the pixels, yeah I am aware of the constraint, the example here was taken merely to demonstrate the difference between the label length and input length for CTC computation.

> I hope that clears your interrogation :)

Yeah. Thanks a lot.
